.github/workflows/todoapp-ci-pipeline.yml

```diff
@@ -16,6 +16,32 @@ jobs:
     name: Build and test
     runs-on: ubuntu-latest
     steps:
-    - uses: actions/checkout@v4
-    - run: docker build --target test --tag todo-app:test .
-    - run: docker run todo-app:test
\ No newline at end of file
+      - uses: actions/checkout@v4
+      - run: docker build --target test --tag todo-app:test .
+      - run: docker run todo-app:test
+
+  push_to_dockerhub:
+    name: Push docker image and deploy
+    runs-on: ubuntu-latest
+    needs: build
+    if: github.ref == 'refs/heads/exercise-9-expanding-ci-pipeline'
+    steps:
+      - uses: actions/checkout@v4
+      - name: Log in to Docker Hub
+        uses: docker/login-action@v3
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
+
+      - name: Build Docker image
+        run: |
+         docker build --target production --tag ${{ secrets.DOCKERHUB_USERNAME }}/todo-app:latest \
+                      --target production --tag ${{ secrets.DOCKERHUB_USERNAME }}/todo-app:${{ github.sha }} .
+
+      - name: Push Docker image
+        run: | 
+          docker push ${{ secrets.DOCKERHUB_USERNAME }}/todo-app:latest
+          docker push ${{ secrets.DOCKERHUB_USERNAME }}/todo-app:${{ github.sha }}
+
+      - name: Deploy to Azure
+        run: curl -v -X POST '${{ secrets.WEBHOOK_URL }}'

```

Comment:
Not important, but there's an interesting outcome in the logs here; your curl response shows in one line:
> {"OperationId":"ab00de6c-2166-4f8d-964c-28521920dad6","TrackingUrl":"https://***todoapp.scm.azurewebsites.net/api/logstream?filter=op:ab00de6c-2166-4f8d-964c-28521920dad6,volatile:false"}

And if we look carefully at the tracking URL, we can make a pretty strong guess that the `***` is probably redacted because of your DOCKERHUB_USERNAME secret. Working backwards from the website URL - it's pretty clear that the username is `clahur`! Now in this case the username could as easily be a variable as a secret, it's not particularly sensitive, so no harm done, but if we had a poor quality password (like some key words related to the project, e.g. `todo-app` or something) this redaction can actually _reveal_ sensitive information ironically!

Nothing I'd actually change, but worth understanding a potential risk